### DISCUSSION TITLE
对 skynet http stream 代码的一些疑问，希望大家指点和探讨一下，谢谢

### DISCUSSION THREAD

**Discussion by davidhenrygao** [cloudwu/skynet#1982] [Q&A]
新项目引入 etcd 组件，watch 监听需要使用 skynet http stream，由于 wiki 上没有相关说明，所以详细看了相关代码，有两点疑问想请教一下：

1. `stream.connected` 赋值为 `nil` 和 `false` 的目的是什么？`stream_length` 和 `stream_read` 两函数的处理似乎不太一致？
2. `stream_all` 使用 `stream_read` 进行 padding 读取，而不是使用 `stream._interface.readall` 进行一次性读取的原因？

这里的说明 https://github.com/cloudwu/skynet/pull/1463#issuecomment-905166350 结合代码，个人理解 `stream.connected = nil` 时应该是表示正常连接关闭的情况，而 `stream.connected = false` 时表示出错的情况。

对于 `stream._interface.read` 函数，对应 http/sockethelper.lua 中 `sockethelper.readfunc` return 的闭包函数，连接正常关闭时，`stream._interface.read()`->`readbytes()`->`socket.read()` 会返回 `false, ret` ，根据代码，应该会直接 error 报错，则 stream_all 最后的读取会报错。不知道我理解是否有误？

`response_stream`代码如下，`mode == "chunked"` 会使用 `stream_chunked` 分片读取，其他情况，若有 "content-length" 头字段表明数据长度，使用 `stream_length(length)` 读取指定长度数据，否则再根据情况不读取 `stream_nobody` 或读取所有剩余数据 `stream_all`。
```lua
function M.response_stream(interface, code, body, header)
	local mode = header["transfer-encoding"]
	if mode then
		if mode ~= "identity" and mode ~= "chunked" then
			error ("Unsupport transfer-encoding")
		end
	end


	local read_func


	if mode == "chunked" then
		read_func = stream_chunked
	else
		-- identity mode
		local length = header["content-length"]
		if length then
			length = tonumber(length)
		end
		if length then
			read_func = stream_length(length)
		elseif code == 204 or code == 304 or code < 200 then
			read_func = stream_nobody
		else
			read_func = stream_all
		end
	end


	-- todo: timeout


	return setmetatable({
		status = code,
		_body = body,
		_interface = interface,
		_reading = read_func,
		header = header,
		connected = true,
	}, stream)
end
``` 

`stream_length 、stream_all 和 stream_read` 代码如下：
```lua
local function stream_length(length)
	return function(stream)
		local body = stream._body
		if body == nil then
			local ret, padding = stream._interface.read()
			if not ret then
				-- disconnected
				body = padding
				stream.connected = false
			else
				body = ret
			end
		end
		local n = #body
		if n >= length then
			stream._reading = stream.close
			stream.connected = nil
			return (body:sub(1,length))
		else
			length = length - n
			stream._body = nil
			if not stream.connected then
				stream._reading = stream.close
			end
			return body
		end
	end
end


local function stream_read(stream)
	local ret, padding = stream._interface.read()
	if ret == "" or not ret then
		stream.connected = nil
		stream:close()
		if padding == "" then
			return
		end
		return padding
	end
	return ret
end


local function stream_all(stream)
	local body = stream._body
	stream._body = nil
	stream._reading = stream_read
	return body
end
``` 

`sockethelper.readfunc` 代码如下：
```lua
function sockethelper.readfunc(fd, pre)
	if pre then
		return preread(fd, pre)
	end
	return function (sz)
		local ret = readbytes(fd, sz)
		if ret then
			return ret
		else
			error(socket_error("read failed fd = " .. fd))
		end
	end
end
``` 

@cloudwu 不清楚云大是否有时间指点一下？
或者其他熟悉相关代码的同学，若我理解有误，也恳请指点和指正一下，谢谢！

**Comment by davidhenrygao** [comment]
另外，因 http stream 的疑问查看相关代码，产生了另外一个疑问：当调用 `socket.readall` 时，`s.read_required = true` 被设置，那么在 SKYNET_SOCKET_TYPE_DATA 事件被处理时，假如数据大小超过 BUFFER_LIMIT，根据代码，应该会 `pause_socket` 屏蔽了底层的 EPOLLIN 事件，那么在连接断开前，都不会被唤醒了？我的理解是这样，不知道是否正确？
```lua
function socket.readall(id)
	local s = socket_pool[id]
	assert(s)
	if not s.connected then
		local r = driver.readall(s.buffer, s.pool)
		return r ~= "" and r
	end
	assert(not s.read_required)
	s.read_required = true
	suspend(s)
	assert(s.connected == false)
	return driver.readall(s.buffer, s.pool)
end
``` 
```lua
-- SKYNET_SOCKET_TYPE_DATA = 1
socket_message[1] = function(id, size, data)
	local s = socket_pool[id]
	if s == nil then
		skynet.error("socket: drop package from " .. id)
		driver.drop(data, size)
		return
	end


	local sz = driver.push(s.buffer, s.pool, data, size)
	local rr = s.read_required
	local rrt = type(rr)
	if rrt == "number" then
		-- read size
		if sz >= rr then
			s.read_required = nil
			if sz > BUFFER_LIMIT then
				pause_socket(s, sz)
			end
			wakeup(s)
		end
	else
		if s.buffer_limit and sz > s.buffer_limit then
			skynet.error(string.format("socket buffer overflow: fd=%d size=%d", id , sz))
			driver.close(id)
			return
		end
		if rrt == "string" then
			-- read line
			if driver.readline(s.buffer,nil,rr) then
				s.read_required = nil
				if sz > BUFFER_LIMIT then
					pause_socket(s, sz)
				end
				wakeup(s)
			end
		elseif sz > BUFFER_LIMIT and not s.pause then
			pause_socket(s, sz)
		end
	end
end
```

**Reply by davidhenrygao** [reply]
我自己写了 client 和 server 服务测试了一下我的疑问，代码就先不贴了，免得又说不清楚。

大概是 client 连接上 server 后，发送一个整数 n 给 server，然后使用 socket.readall(id) 进行阻塞读取； server 收到请求后，返回 n KB字节数据给 client。当 client 发送 n = 1024 时，打印显示会进入 `pause_socket` 函数，底层会屏蔽了 EPOLLIN 事件。server 发送完数据后，socket.close 关闭连接，tcpdump 显示有发送 FIN 报文到 client，但此时 client 会一直阻塞，因为不会有 EPOLLIN 事件，也不会有 EPOLLHUP 事件。

@cloudwu 云大能不能顺道解答一下？这里调用 `pause_socket` 的目的是什么？
https://github.com/cloudwu/skynet/blob/8900360fab6b449d63ab9d5efcac2d4088d5e060/lualib/skynet/socket.lua#L99-L101

**Reply by cloudwu** [reply]
这里 `BUFFER_LIMIT` 其实是一个防呆设定，来源于早期对 readall 设计的考虑不周。readall 早于 pause 机制实现。

因为 readall 的语义是读 fd 直到对端关闭，中间没有干涉的手段。如果对端不友好，那么只要一直发送数据而不关闭，最终接收方内存一定会耗尽。这会是一个非常简单的攻击手段。这也是为什么前面提到不建议使用 readall 的原因。

**Reply by davidhenrygao** [reply]
readall 设计上确实欠佳，但目前使用 `pause_socket` 进行约束，感觉对尚存的使用者不太友好，因为 pause 后就无法恢复了。

在目前 readall 函数向前兼容并存在的情况下，我的一个想法是，增加一个可选参数 limit，默认值为 BUFFER_LIMIIT，超过 limit 值就直接报错，因为目前 pause 后其实也无法恢复。

**Comment by cloudwu** [comment]
我觉得这里给的 interface 是个未实现的功能组，应该是依赖数量越少越好。如果可以不依赖 readall 就不应该依赖。readall 本身也是一个不太正常的功能，我认为应该回避使用。

另外 steam 既然是流处理，也不应该把整个流的数据全部读完 (all) 再处理。

**Reply by davidhenrygao** [reply]
大致上明白这里的设计了。

我上面说得比较啰嗦，可能说的一个问题你没留意到：`response_stream` 函数处理，当模式不为 `chunked`，并且没有 `content-length` header，那么会使用 `sream_all` 函数处理，但目前有个问题，就是当 web 服务器正常发送完 http 响应数据并关闭连接，客户端循环调用 stream() 读取数据，最后一次会 error 报错，因为目前 `interface.read` 的具体实现使然。个人预期是正常的读取完，不会报错。

我简单写了个服务连接 skynet 自带的 simpleweb 服务测试了一下：
```lua
local skynet = require "skynet"
local Httpc = require "http/httpc"

skynet.start(function()
    Httpc.dns()
    skynet.error("Http stream test begin.")
    local ok, stream = pcall(Httpc.request_stream, "POST", "127.0.0.1:8001",
        "/test/stream", nil, {}, "test")
    if not ok then
        skynet.error(string.format("httpc request_stream failed: %s\n", stream))
    else
        local data = ""
        local padding
        while true do
            ok, padding = pcall(stream)
            if ok then
                if padding then
                    data = data .. padding
                else
                    skynet.error(string.format("stream end, data: \n%s", data))
                    break
                end
            else
                skynet.error(string.format("stream call error: %s, cur data: \n%s", padding, data))
                break
            end
        end
    end
    skynet.error("Http stream test end.")
    skynet.exit()
end)
``` 
输出：
```text
[:01000025] Http stream test begin.
[:01000010] 127.0.0.1:53346 connected, pass it to agent :01000011
[:01000025] stream end, data:
host: 127.0.0.1:8001
path: /test/stream
-----header----
content-length = 4
host = 127.0.0.1:8001
-----body----
test
[:01000025] Http stream test end.
``` 
修改 lualib/http/httpd.lua，故意去除 `content-length` 信息，再测试：
```lua
116 local function writeall(writefunc, statuscode, bodyfunc, header)
       ......
130         local t = type(bodyfunc)
131         if t == "string" then
132                 --writefunc(string.format("content-length: %d\r\n\r\n", #bodyfunc))
133                 writefunc("\r\n")
134                 writefunc(bodyfunc)
135         elseif t == "function" then
```
输出：
```text
[:01000025] Http stream test begin.
[:01000010] 127.0.0.1:54314 connected, pass it to agent :01000011
[:01000025] stream call error: [Socket Error] : read failed fd = 9, cur data:
host: 127.0.0.1:8001
path: /test/stream
-----header----
content-length = 4
host = 127.0.0.1:8001
-----body----
test
[:01000025] Http stream test end.
```
故意去除 `content-length` 信息前，`response_stream` 使用 `stream_length`，去除后使用的是 `stream_all`。

**Reply by cloudwu** [reply]
http 1.0 的协议现在已经很少用了，所以这里处理比较粗糙。

你可以改一下 `stream_read` 函数提个 pr ：

在调用 interface.read() 后，如果发现对端断开，应该设置为 `stream._reading = stream.close` ，走 close 流程。

**Reply by davidhenrygao** [reply]
> 你可以改一下 stream_read 函数提个 pr ：
在调用 interface.read() 后，如果发现对端断开，应该设置为 stream._reading = stream.close ，走 close 流程。

我没太明白你这里说的。对端断开，现在调用 interface.read() 会直接报错，无法在调用后判断。

问题其实在于 `interface.read` 的实现上，对应 lualib/http/sockethelper.lua 代码：
```lua
function sockethelper.readfunc(fd, pre)
	if pre then
		return preread(fd, pre)
	end
	return function (sz)
		local ret = readbytes(fd, sz)
		if ret then
			return ret
		else
			error(socket_error("read failed fd = " .. fd))
		end
	end
end
```
`readbytes` 其实就是 `socket.read` 函数，连接断开时，第一个返回值为 false，但上面代码直接 error 报错了，导致 `stream_read` 函数的 interface.read() 调用也直接报错。

之前用过 websocket ，那边也会用 sockethelper 这个库，能理解直接报错的设计目的，websocket 那边必定会传入 sz 参数进行调用，并且默认只返回 sz 大小的 data。

**或许说问题出现在对 interface.read 函数的使用上，是否可以不传入参数 sz？函数返回值是什么？个人感觉 httpc 和 websocket 两处在对 interface.read 函数的使用上并不一致**

若要兼容两者，是不是可以这样修改一下（`preread` 函数也需同步修改）？但感觉有点怪异。
```lua
function sockethelper.readfunc(fd, pre)
	if pre then
		return preread(fd, pre)
	end
	return function (sz)
		local ret, padding = readbytes(fd, sz)
		if sz then
			if ret then
				return ret
			else
				error(socket_error("read failed fd = " .. fd))
			end
		else
			return ret, padding
		end
	end
end
```

**Reply by cloudwu** [reply]
socket.read 当不传 sz 的时候，表示尽可能的把底层 buffer 中的数据返回，但至少返回一个字节。它用 false 表示对端已经关闭。

我现在觉得，使用 sz 参数的不同类型 （nil or number) 来区分两种不同的用法是不好的。上面的改法我觉得可以，但的确又继续在使用不好的用法。所以，是不是可以考虑 stream 的实现上也改为给一个固定的 sz 一段段读？避免 sz 为 nil 的情况。

**Reply by davidhenrygao** [reply]
> 使用 sz 参数的不同类型 （nil or number) 来区分两种不同的用法是不好的。

我也是这样觉得的。

> 所以，是不是可以考虑 stream 的实现上也改为给一个固定的 sz 一段段读？避免 sz 为 nil 的情况。

但这个想法，我没有想到具体的实现方向。因为目前报错是发生在 `stream_all` 的情况，此时并不知道需要读多少数据。

我的一个想法是，interface 是不是可以增加一个 `readany` 接口，来取代不传 sz 参数的用法，原来的 read 必须传 sz 参数。
可能这跟 `socket.read` 的使用上产生了分歧，但其实我觉得不好的用法的根源或许就是 `socket.read`  原有的设计？

**Reply by cloudwu** [reply]
steam 的设计，那个读取函数是循环调用的，所以随便设一个数字应该就可以工作。

readany 有点像是画蛇添足，因为传统 socket api 都不需要对应 api 。

**Reply by cloudwu** [reply]
另外，我依稀记得对于 http 1.0 协议的考虑还有一点。当服务器不提供内容长度时，主动关闭连接的责任最好放在客户端而不是服务端。即，客户端觉得自己读完了所有内容就主动关闭连接，而不应该等到发现服务端关闭连接。

这本身也可能导致大量的 tcp TIMEWAIT。

**Reply by davidhenrygao** [reply]
> steam 的设计，那个读取函数是循环调用的，所以随便设一个数字应该就可以工作。

我还是没太明白 😂 

例如这个数字设为 10，那会跟正常要读取 sz = 10 冲突吧？设为 -1 的话，那其实跟判断 nil 没区别了。
```lua
function sockethelper.readfunc(fd, pre)
	if pre then
		return preread(fd, pre)
	end
	return function (sz)
		local ret = readbytes(fd, sz)
		if ret then
			return ret
		else
			error(socket_error("read failed fd = " .. fd))
		end
	end
end
``` 
这里我理解的设计意图就是：一定要读取到 sz 字节，不然就报错，连接断开也报错，websocket 那边也全是按照这个逻辑使用的，使用起来也方便。

但 `stream_all` 函数就是不知道要读多少，假如设 sz 为 10 一段段读，假设服务器发了 36 字节数据过来，那最后一次读，readbytes 返回 flase, padding，padding 是最后 6 字节内容，也是会报错的吧。

难度云大你是想把 stream、websocket 和 sockethelper 相关地方都统一修改？只修改 stream 好像无法处理。

**Reply by cloudwu** [reply]
我搞错了。目前看起来需要对 sz 为 nil 单独处理，即上面的改法。就是让其不报错。当然，如果接受报错分支正确处理 error ，我想也可以把事情做完。

**Reply by davidhenrygao** [reply]
> 另外，我依稀记得对于 http 1.0 协议的考虑还有一点。当服务器不提供内容长度时，主动关闭连接的责任最好放在客户端而不是服务端。即，客户端觉得自己读完了所有内容就主动关闭连接，而不应该等到发现服务端关闭连接。
> 
> 这本身也可能导致大量的 tcp TIMEWAIT。

不提供内容长度，感觉客户端也无法判断是否读完吧，例如内容格式是 json，那么只能每次读取完就去尝试判断是否是合法的 json 数据，感觉是不是做了多余的判断？

可能导致大量的 tcp TIMEWAIT：是指使用 skynet 的 httpc 作为客户端访问 http 服务器，skynet 作为客户端主动关闭连接，可能产生大量 tcp TIMEWAIT 的问题？

**Reply by cloudwu** [reply]
我指 http 服务端发送完数据就断开，可能产生 timewait ；反之 http server 不主动断开，而是 client 认为自己获得了足够数据后断开，则不会。所以我倾向于避免使用连接断开作为明确的流结束的信号。

**Reply by cloudwu** [reply]
现在已经几乎不存在还在使用 http 1.0 协议的场所了，即内容长度是必填项。所以我觉得这里的问题不是什么大问题。

**Reply by davidhenrygao** [reply]
> 当然，如果接受报错分支正确处理 error ，我想也可以把事情做完。

这是指在 else 分支再判断是否报 error ？还是使用者外层报错处理？使用者无法处理吧？

**Reply by davidhenrygao** [reply]
> 现在已经几乎不存在还在使用 http 1.0 协议的场所了，即内容长度是必填项。所以我觉得这里的问题不是什么大问题。

的确是一些几乎不会遇到的问题，一开始就是代码里面看到了，感觉没有内容长度的时候，是会报错的，所以在这里问问看。

那这个还要改吗？

**Reply by davidhenrygao** [reply]
那其实是不是就不要兼容 http1.0 没有长度头的情况了？不要 `stream_all` ，直接 close 连接，不读 body 算了。

**Comment by cloudwu** [comment]
`stream.connected` 为 nil 时，这个 http 流是正常结束的，false 指某个环节的数据不正常。

**Comment by cloudwu** [comment]
我想了一下，目前解决这个问题伤害最小的改法还是加一个 interface.readany ，然后让 stream api 使用它。但以后在别的模块使用 readany 应该斟酌。

这样可以最小程度影响已有代码的行为，同时解决这个指定问题。

如果想兼容再好一些，当 readany 不存在时，转发到 read 更好。

**Reply by cloudwu** [reply]
在处理这些非常见情况的问题时，我倾向于增加一些代码处理特例，避免修改已有代码的行为。

**Reply by davidhenrygao** [reply]
确实，刚刚在 tls 那边看到有不带 sz 的调用，判断 sz 为 nil 的改法会有影响。
